File: python_test_crate/tests/21_actions_basic.py
from typing import Any, Optional, List, Dict, Callable

class ActionsTest:
    def __init__(self):
        self._state_stack = []
        self._state_context = {}
        self._return_value = None
        self.log = ""
        self._state = "Ready"
        self._enter()

    def _transition(self, target_state, exit_args = None, enter_args = None):
        if exit_args:
            self._exit(*exit_args)
        else:
            self._exit()
        self._state = target_state
        if enter_args:
            self._enter(*enter_args)
        else:
            self._enter()

    def _change_state(self, target_state):
        self._state = target_state

    def _dispatch_event(self, event, *args):
        handler_name = f"_s_{self._state}_{event}"
        handler = getattr(self, handler_name, None)
        if handler:
            return handler(*args)

    def _enter(self, *args):
        # No enter handlers
        pass

    def _exit(self, *args):
        # No exit handlers
        pass

    def process(self, value: int) -> int:
        self._return_value = None
        self._dispatch_event("process", value)
        return self._return_value

    def get_log(self) -> str:
        self._return_value = None
        self._dispatch_event("get_log")
        return self._return_value

    def _s_Ready_process(self, value: int) -> int:
        self.__log_event("start")
        self.__validate_positive(value)
        self.__log_event("valid")
        result = value * 2
        self.__log_event("done")
        self._return_value = result
        return

    def _s_Ready_get_log(self) -> str:
        self._return_value = self.log
        return

    def __log_event(self, msg: str):
        self.log = self.log + msg + ";"

    def __validate_positive(self, n: int):
        if n < 0:
            raise ValueError(f"Value must be positive: {n}")


def main():
    print("=== Test 21: Actions Basic ===")
    s = ActionsTest()

    # Test 1: Actions are called correctly
    result = s.process(5)
    assert result == 10, f"Expected 10, got {result}"
    print(f"1. process(5) = {result}")

    # Test 2: Log shows action calls
    log = s.get_log()
    assert "start" in log, f"Missing 'start' in log: {log}"
    assert "valid" in log, f"Missing 'valid' in log: {log}"
    assert "done" in log, f"Missing 'done' in log: {log}"
    print(f"2. Log: {log}")

    # Test 3: Action with validation
    try:
        s.process(-1)
        assert False, "Should have raised ValueError"
    except ValueError as e:
        print(f"3. Validation caught: {e}")

    print("PASS: Actions basic works correctly")

if __name__ == '__main__':
    main()

File: typescript_test_crate/tests/21_actions_basic.ts
class ActionsTest {
    private _state: string;
    private _state_stack: Array<any>;
    private _state_context: Record<string, any>;
    private _return_value: any;
    private log: string = "";

    constructor() {
        this._state_stack = [];
        this._state_context = {  };
        this._return_value = null;
        this.log = "";
        this._state = "Ready";
        this._enter();
    }

    private _transition(target_state: string, exit_args: any = null, enter_args: any = null) {
        if (exit_args) {
            this._exit(...exit_args);
        } else {
            this._exit();
        }
        this._state = target_state;
        if (enter_args) {
            this._enter(...enter_args);
        } else {
            this._enter();
        }
    }

    private _change_state(target_state: string) {
        this._state = target_state;
    }

    private _dispatch_event(event: string, ...args: any[]) {
        const handler_name = `_s_${this._state}_${event}`;
        const handler = (this as any)[handler_name];
        if (handler) {
            return handler.apply(this, args);
        }
    }

    private _enter(...args: any[]) {
        // No enter handlers
    }

    private _exit(...args: any[]) {
        // No exit handlers
    }

    public process(value: number): number {
        this._return_value = null
        this._dispatch_event("process", value)
        return this._return_value
    }

    public get_log(): string {
        this._return_value = null
        this._dispatch_event("get_log")
        return this._return_value
    }

    private _s_Ready_get_log() {
        this._return_value = this.log;
        return this._return_value;;
    }

    private _s_Ready_process(value: number) {
        this.__log_event("start");
        this.__validate_positive(value);
        this.__log_event("valid");
        const result = value * 2;
        this.__log_event("done");
        this._return_value = result;
        return this._return_value;;
    }

    private __log_event(msg: string) {
                    this.log = this.log + msg + ";";
    }

    private __validate_positive(n: number) {
                    if (n < 0) {
                        throw new Error(`Value must be positive: ${n}`);
                    }
    }
}


function main(): void {
    console.log("=== Test 21: Actions Basic (TypeScript) ===");
    const s = new ActionsTest();

    // Test 1: Actions are called correctly
    const result = s.process(5);
    if (result !== 10) {
        throw new Error(`Expected 10, got ${result}`);
    }
    console.log(`1. process(5) = ${result}`);

    // Test 2: Log shows action calls
    const log = s.get_log();
    if (!log.includes("start")) {
        throw new Error(`Missing 'start' in log: ${log}`);
    }
    if (!log.includes("valid")) {
        throw new Error(`Missing 'valid' in log: ${log}`);
    }
    if (!log.includes("done")) {
        throw new Error(`Missing 'done' in log: ${log}`);
    }
    console.log(`2. Log: ${log}`);

    // Test 3: Action with validation
    try {
        s.process(-1);
        throw new Error("Should have thrown Error");
    } catch (e) {
        if (e instanceof Error && e.message.includes("positive")) {
            console.log(`3. Validation caught: ${e.message}`);
        } else {
            throw e;
        }
    }

    console.log("PASS: Actions basic works correctly");
}

main();

